### PR TITLE
Ci setup

### PR DESCRIPTION
**Pull request by lucasberent**
WIP created to trigger github ci
do not merge

### CODE REVIEW

File: src/DecodingSimulator.cpp [hunk 1]
```diff
@@ -7,10 +7,12 @@
 #include "UFDecoder.hpp"
 
 std::string generateOutFileName(const std::string& filepath) {
-    auto               t  = std::time(nullptr);
-    auto               tm = *std::localtime(&t);
+    std::time_t now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    char buf[16] = { 0 };
+    std::strftime(buf, sizeof(buf), "%Y-%m-%d", std::localtime(&now));
```
[the remark this answers is not in the document]
Comment (lucasberent) [reply src/DecodingSimulator.cpp:12]: added comment, currently not relevant.

File: .cmake-format.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+format:
+  line_width: 100
+  keyword_case: "upper"
+  autosort: true
```
Comment (burgholzer) [inline .cmake-format.yaml:4]: there is a newline missing at the end of this file.

File: .flake8 [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    .idea,
+    .eggs,
+    __pycache__,
+    .tox,
+    docs/source/conf.py,
+    build,
+    .nox,
+    venv,
+    .venv
+application-import-names = mqt.qecc
```
Comment (burgholzer) [inline .flake8:17]: there is a newline missing at the end of this file.

File: .github/ISSUE_TEMPLATE/bug-report.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  - type: textarea
+    attributes:
+      label: How to Reproduce
+      description: Please provide steps to reproduce this bug.
+      placeholder: |
+        1. Get package from '...'
+        2. Then run '...'
+        3. An error occurs.
+    validations:
+      required: true
```
Comment (burgholzer) [inline .github/ISSUE_TEMPLATE/bug-report.yml:63]: there is a newline missing at the end of this file.

File: .github/ISSUE_TEMPLATE/feature-request.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  - type: textarea
+    attributes:
+      label: Describe the solution you'd like
+      description: >
+        Clear and concise description of what you want to happen.
+      placeholder: >-
+        When I do X, I want to achieve Y in a situation when Z.
+    validations:
+      required: true
```
Comment (burgholzer) [inline .github/ISSUE_TEMPLATE/feature-request.yml:35]: there is a newline missing at the end of this file.

File: .github/workflows/python-ci.yml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      - if: runner.os != 'Linux'
+        name: Test on 3.10
+        run: nox -s tests-3.10
+      # Run the coverage job on Python 3.10 and upload the coverage data.
+      - if: runner.os == 'Linux'
+        name: Test and Coverage on 3.10
+        run: nox -s coverage-3.10 -- --cov-report=xml
+      - if: runner.os == 'Linux'
+        name: Upload Coverage to Codecov
+        uses: codecov/codecov-action@v3.1.1
```
Comment (burgholzer) [inline .github/workflows/python-ci.yml:72]: there is a newline missing at the end of this file.

File: .lgmt.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    after_prepare:
+      - export PATH=~/.local/bin:$PATH
+      - pip3 install cmake --user
+      - cmake --version
+    configure:
+      command:
+        - cmake -S . -B build -DBUILD_QECC_TESTS=ON -DBINDINGS=ON
+    index:
+      build_command:
+        - cmake --build build --parallel 4
```
Comment (burgholzer) [inline .lgmt.yaml:20]: this file can be removed as LGTM is deprecated and will be shut down at the end of the year

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        additional_dependencies: [pyyaml]
+      - id: cmake-lint
+        additional_dependencies: [pyyaml]
+
+  # Format configuration files with prettier
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: "v3.0.0-alpha.1"
+    hooks:
+      - id: prettier
+        types_or: [yaml, markdown, html, css, scss, javascript, json]
```
Comment (burgholzer) [inline .pre-commit-config.yaml:140]: there is a newline missing at the end of this file.

File: include/Code.hpp [hunk 9]
```diff
@@ -200,40 +192,41 @@ class Code {
...
-        } else if (err.size() > this->getN()) {
+        }
+        if (err.size() > this->getN()) {
             std::vector<bool> xerr;
             xerr.reserve(getN());
             std::vector<bool> zerr;
             zerr.reserve(getN());
-            std::move(err.begin(), err.begin() + getN(), std::back_inserter(xerr));
-            std::move(err.begin() + getN(), err.end(), std::back_inserter(zerr));
+            std::move(err.begin(), err.begin() + static_cast<int64_t>(getN()), std::back_inserter(xerr));
```
Comment (burgholzer) [inline include/Code.hpp:201]:
```suggestion
            std::move(err.begin(), err.begin() + static_cast<std::int64_t>(getN()), std::back_inserter(xerr));
```

File: include/Code.hpp [hunk 10]
```diff
@@ -200,40 +192,41 @@ class Code {
...
+            std::move(err.begin() + static_cast<int64_t>(getN()), err.end(), std::back_inserter(zerr));
```
Comment (burgholzer) [inline include/Code.hpp:202]:
```suggestion
            std::move(err.begin() + static_cast<std::int64_t>(getN()), err.end(), std::back_inserter(zerr));
```

File: include/Decoder.hpp [hunk 11]
```diff
@@ -15,49 +15,49 @@
...
+    SingleQubitRandom
 };
 
 [[maybe_unused]] static GrowthVariant growthVariantFromString(const std::string& architecture) {
     if (architecture == "ALL_COMPONENTS" || architecture == "0") {
-        return GrowthVariant::ALL_COMPONENTS;
-    } else if (architecture == "INVALID_COMPONENTS" || architecture == "1") {
-        return GrowthVariant::INVALID_COMPONENTS;
+        return GrowthVariant::AllComponents;
+    } else if (architecture == "INVALID_COMPONENTS" || architecture == "1") { // NOLINT(readability-else-after-return)
```
Comment (burgholzer) [inline include/Decoder.hpp:28]: Why not just replace these `else if's with just if's as suggested by the warning?

File: include/TreeNode.hpp [hunk 12]
```diff
@@ -22,40 +23,37 @@ class TreeNode {
...
+        auto* parent = node->parent;
         while (parent != nullptr && parent->parent != nullptr) {
-            node = parent;
+            node   = parent;
             parent = parent->parent;
         }
         if (parent == nullptr) {
             return node;
-        } else {
+        } else { // NOLINT(readability-else-after-return)
```
Comment (burgholzer) [inline include/TreeNode.hpp:47]: don't ignore that warning. just replace the else if with a regular if

File: include/TreeNode.hpp [hunk 13]
```diff
@@ -22,40 +23,37 @@ class TreeNode {
...
             return parent;
         }
     }
     /*
      * Merge two trees with given roots
      */
-    static void Union(TreeNode* tree1, TreeNode* tree2) {
-        auto        root1 = Find(tree1);
-        auto        root2 = Find(tree2);
+    static void Union(TreeNode* tree1, TreeNode* tree2) { // NOLINT(readability-identifier-naming)
```
Comment (burgholzer) [inline include/TreeNode.hpp:54]: why not rename it to `union`?

File: mqt/qecc/py.typed [hunk 14]
```diff
@@ -0,0 +1,2 @@
+# Instruct type checkers to look for inline type annotations in this package.
+# See PEP 561.
```
Comment (burgholzer) [inline mqt/qecc/py.typed:2]: there is a newline missing at the end of this file.

File: noxfile.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        session.install("-e", ".[docs]")
+    session.chdir("docs")
+    session.run("sphinx-build", "-M", "html", "source", "_build")
+
+    if session.posargs:
+        if "serve" in session.posargs:
+            print("Launching docs at http://localhost:8000/ - use Ctrl-C to quit")
+            session.run("python", "-m", "http.server", "8000", "-d", "_build/html")
+        else:
+            print("Unsupported argument to docs")
```
Comment (burgholzer) [inline noxfile.py:109]: there is a newline missing at the end of this file.

File: scripts/numerics/visualize/visualizeDecodingRuntime.py [hunk 16]
```diff
@@ -49,7 +51,6 @@ def runtime():
     yfinal = []
     cols = []
     colors = mcolors.BASE_COLORS
```
Comment (burgholzer) [inline scripts/numerics/visualize/visualizeDecodingRuntime.py:53]: This is apparently defined a second time down below. Consider removing the definition below. (see GitHub CodeQL warnings)

File: scripts/numerics/visualize/visualizeWordErrRate.py [hunk 17]
```diff
@@ -13,6 +13,7 @@ def wer():
 
     f = open(inputFilename)
```
Comment (burgholzer) [inline scripts/numerics/visualize/visualizeWordErrRate.py:14]: there are still some GutHub CodeQL warnings in this file

File: src/UFDecoder.cpp [hunk 18]
```diff
@@ -17,21 +17,21 @@
  * @param syndrome
  */
 void UFDecoder::decode(const gf2Vec& syndrome) {
-    if (syndrome.size() > this->getCode()->getHz()->pcm->size()) {
+    if (syndrome.size() > this->getCode()->gethZ()->pcm->size()) {
         std::vector<bool> xSyndr;
         std::vector<bool> zSyndr;
-        auto              mid = syndrome.begin() + (syndrome.size() / 2U);
+        auto              mid = syndrome.begin() + (static_cast<int64_t>(syndrome.size()) / 2U);
```
Comment (burgholzer) [inline src/UFDecoder.cpp:23]:
```suggestion
        auto              mid = syndrome.begin() + (static_cast<std::int64_t>(syndrome.size()) / 2U);
```

**Comment by lucasberent** [comment]
For now windows wheels are excluded, see #4.

**Review by burgholzer — CHANGES_REQUESTED** [review]
I just quickly went through all the changes. Overall this looks good and it is great to have all this set up.

You can find some comments inline. (some of those could have been fixed automatically by running the pre-commit checks; `pre-commit run --all-files`)
There are still a couple of CodeQL warnings that should be rather easy to address.

File: include/TreeNode.hpp (hunk 13 above)
Comment (lucasberent) [reply include/TreeNode.hpp:54]: its a reserved keyword.  thats why i'd like to keep Find and Union for consistency.

Comment (burgholzer) [reply include/TreeNode.hpp:53]: Makes sense, didn't think about that 👍🏻

**Review by burgholzer — APPROVED** [review]
LGTM 👍🏼 
We can always issue a bugfix if something should turn out to not work at all.

**Pull request merged**